File: Leetcode/199.二叉树的右视图.cpp
/*
 * @lc app=leetcode.cn id=199 lang=cpp
 *
 * [199] 二叉树的右视图
 */
#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;
/*2021-09-14-10-45*/
// ? 思路解析：
/*
 *
 *
 !
 !
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right)
        : val(x), left(left), right(right) {}
};

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
   public:
    vector<int> rightSideView(TreeNode *root) {
        vector<int> ans;
        if (root == NULL) return ans;
        queue<TreeNode *> que;

        if (root != NULL) que.push(root);
        while (!que.empty()) {
            int size = que.size();
            for (int i = 0; i < size; i++) {
                TreeNode *node = que.front();
                que.pop();

                if (i == (size - 1)) ans.push_back(node->val);
                
                if (node->left) que.push(node->left);
                if (node->right) que.push(node->right);
            }
        }
        return ans;
    }
};
// @lc code=end

File: Leetcode/199.二叉树的右视图.go
/*
 * @lc app=leetcode.cn id=199 lang=golang
 *
 * [199] 二叉树的右视图
 */
package leetcode

/* --- 2022-01-16-13-19 --- */

// https://github.com/fengwei2002/Algorithm
// solution link:
// https://leetcode-cn.com/problems/binary-tree-right-side-view/solution/lc199-fengwei2002-by-kycu-n845/

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rightSideView(root *TreeNode) []int {
	que := make([]*TreeNode, 0)
	ans := make([]int, 0)
	if root == nil {
		return ans
	}

	que = append(que, root)
	for len(que) > 0 {
		length := len(que)

		for i := 0; i < length; i++ {
			t := que[0]
			que = que[1:]

			if t.Left != nil {
				que = append(que, t.Left)
			}
			if t.Right != nil {
				que = append(que, t.Right)
			}
			if i == length-1 {
				ans = append(ans, t.Val)
			}
		}
	}
	return ans
}

// @lc code=end
